lib/cli/src/Cardano/CLI.hs

```diff
@@ -143,6 +146,15 @@ parseAllArgsWith cli args option = do
     getAllArgsOrExit :: Arguments -> Option -> IO (NE.NonEmpty String)
     getAllArgsOrExit = getAllArgsOrExitWith cli
 
+parseOptionalArg :: FromText a => Arguments -> Option -> IO (Maybe a)
+parseOptionalArg args option =
+    case fromText . T.pack <$> args `getArg` option of
+        Nothing -> pure Nothing
+        Just (Right a) -> pure $ pure a
+        Just (Left e) -> do
+            putErrLn $ T.pack $ getTextDecodingError e
+            exitFailure
```

Comment:
It would be nice if we could print help (the `Docopt`) to the console after the failure instead (what `getArgOrExitWith` does)

Reply:
@akegalj wrote:

> It would be nice if we could print help (the `Docopt`) to the console after the failure instead (what `getArgOrExitWith` does)

Thanks for this suggestion.

Actually, I can't help wondering if it wouldn't be nicer to print out something _even more specific_. Instead of printing out the entire help string whenever there's a problem, perhaps we could print out **just the specific part that is relevant**.

For example:

```
$ cardano-wallet address list --state on-fire 959b511ffa8213da57443226e3f9e230f956c446

Error: invalid address state: it should be either "used" nor "unused"

Usage: cardano-wallet address list [--port=INT] [--state=STRING] <wallet-id>
```
See here for a longer discussion: https://github.com/input-output-hk/cardano-wallet/issues/371

I'm not sure if it's easy to do this with `docopt`. (Perhaps switching `optparse-applicative` would make this easier.)

Reply:
@akegalj Given that we can't (currently) print help that is specific to just one command, is it still your preference to print out the whole help string on failure? If so, I can add this in.

Reply:
@akegalj We currently have this output when a user supplies an invalid value for `--state`:

```
$ cardano-wallet address list --state on-fire 959b511ffa8213da57443226e3f9e230f956c446
Unable to decode address state: it's neither "used" nor "unused"
```

Reply:
I am fine with the user output in this case. 

And in general I am not sure that printing out the whole help as we usually do it atm is a user friendly thing to do. I would rather say "See 'cardano-wallet bla --help' for more details instead if you are interested" because if the whole help is printed then that usually hides the relevant error message.

I have proposed to print out help in this case just so that we can be consistent with the rest of the code - and so that we don't mix multiple styles of error handling in our CLI.